File: CSV2HTML_by_Java/csv2html/Writer.java
package csv2html;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.List;
import utility.StringUtility;

/**
 * ライタ：CSV情報のテーブルをHTMLページとして書き出す。
 * @author 北村拓海
 * @since 2024/12/01
 * @version 1.0
 */
public class Writer extends IO {
	

    /**
     * ライタのコンストラクタ。
     * @param aTable テーブル
     */
    public Writer(Table aTable) {
        super(aTable);
    }

    /**
     * HTMLページを基にするテーブルからインデックスファイル(index.html)に書き出す。
     */
    public void perform() {
        try {
            Attributes attributes = this.attributes();
            String fileStringOfHTML = attributes.baseDirectory() + attributes.indexHTML();
            File aFile = new File(fileStringOfHTML);
            FileOutputStream outputStream = new FileOutputStream(aFile);
            OutputStreamWriter outputWriter = new OutputStreamWriter(outputStream, StringUtility.encodingSymbol());
            BufferedWriter aWriter = new BufferedWriter(outputWriter);

            this.writeHeaderOn(aWriter);
            this.writeTableBodyOn(aWriter);
            this.writeFooterOn(aWriter);

            aWriter.close();
        } catch (UnsupportedEncodingException | FileNotFoundException anException) {
            anException.printStackTrace();
        } catch (IOException anException) {
            anException.printStackTrace();
        }
    }

    /**
     * 属性リストを書き出す。
     * @param aWriter ライタ
     */
    public void writeAttributesOn(BufferedWriter aWriter) throws IOException {
        aWriter.write("<tr>");
        aWriter.newLine();
        
        for (String attribute : this.table().attributes().names()) {
            aWriter.write("<td class=\"center-pink\"><strong>" + attribute + "</strong></td>");
            aWriter.newLine();
        }
        
        aWriter.write("</tr>");
        aWriter.newLine();
    }

    /**
     * フッタを書き出す。
     * @param aWriter ライタ
     */
    public void writeFooterOn(BufferedWriter aWriter) throws IOException {
        aWriter.write("</table>");
        aWriter.newLine();
        aWriter.write("</body>");
        aWriter.newLine();
        aWriter.write("</html>");
        aWriter.newLine();
    }

    /**
     * ヘッダを書き出す。
     * @param aWriter ライタ
     */
    public void writeHeaderOn(BufferedWriter aWriter) throws IOException {
        aWriter.write("<!DOCTYPE html>");
        aWriter.newLine();
        aWriter.write("<html>");
        aWriter.newLine();
        aWriter.write("<head>");
        aWriter.write("<meta charset=\"UTF-8\">");
        aWriter.write("<style>");
        aWriter.write("table { border-collapse: collapse; width: 100%; }");
        aWriter.write("th, td { border: 1px solid black; padding: 8px; text-align: left; }");
        aWriter.write("th { background-color: #ffddbb; }");
        aWriter.write(".even-row { background-color: #ccffcc; }");
        aWriter.write(".odd-row { background-color: #ffddbb; }");
        aWriter.write("img { display: block; margin: auto; }");
        aWriter.write("</style>");
        aWriter.write("</head>");
        aWriter.newLine();
        aWriter.write("<body>");
        aWriter.newLine();
        aWriter.write("<table>");
        aWriter.newLine();
    }

    /**
     * ボディを書き出す。
     * @param aWriter ライタ
     */
    public void writeTableBodyOn(BufferedWriter aWriter) throws IOException {
        this.writeAttributesOn(aWriter); // 属性リスト（ヘッダー）を書き出す
        this.writeTuplesOn(aWriter); // データ行（タプル群）を書き出す
    }

    /**
     * タプル群を書き出す。
     * @param aWriter ライタ
     */
    private void writeTuplesOn(BufferedWriter aWriter) throws IOException {
        aWriter.write("<h1>総理大臣</h1>");
        int index = 0;
        for (Tuple aTuple : this.table().tuples()) {
            System.out.println("aTuple: " + aTuple);
            System.out.println("");
            aWriter.write("<tr>");
            aWriter.newLine();
            
            String rowClass = (index % 2 == 0) ? "even-row" : "odd-row";
            for (String aString : aTuple.values()) {
                aWriter.write("<td class=\"" + rowClass + "\">");
                
                // エスケープされたHTMLタグを元に戻して処理
                String unescapedString = aString
                    .replace("&lt;", "<")
                    .replace("&gt;", ">")
                    .replace("&quot;", "\"")
                    .replace("&amp;", "&");
                    
                if (unescapedString.startsWith("<a href='images/") && unescapedString.endsWith("</a>")) {
                    // 画像タグをそのまま出力
                    aWriter.write(unescapedString);
                } else if (unescapedString.matches(".*\\d+\\.jpg$")) {
                    String imageNumber = unescapedString.replaceAll("[^0-9]", "");
                    aWriter.write("<a href=\"images/" + imageNumber + ".jpg\">");
                    aWriter.write("<img src=\"thumbnails/" + imageNumber + ".jpg\" " +
                                "width=\"32\" height=\"32\" " +
                                "alt=\"Image " + imageNumber + "\">");
                    aWriter.write("</a>");
                } else {
                    aWriter.write(IO.htmlCanonicalString(unescapedString));
                }
                
                aWriter.write("</td>");
                aWriter.newLine();
            }
            aWriter.write("</tr>");
            aWriter.newLine();
            index++;
        }
    }
}

File: CSV2HTML_by_Python/csv2html/writer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""CSV2HTML：総理大臣と徳川幕府の情報「CSV」からWebページ「HTML」を生成。"""
__author__ = 'OGAWA YUITO'
__version__ = '1.0.7'
__date__ = '2024/12/20 (Created: 2024/12/20)'

import datetime
import os

from csv2html.io import IO

# pylint: disable=R0201
# R0201: Method could be a function (no-self-use)

class Writer(IO):
	"""ライタ：情報のテーブルをHTMLページとして書き出す。"""

	def __init__(self, output_table):
		"""ライタのコンストラクタ。HTMLページを基にするテーブルを受け取る。"""

		super().__init__(output_table)
		(lambda x: x)(output_table) # NOP

	def perform(self):
		"""HTMLページを基にするテーブルから、インデックスファイル(index_html)に書き出す。"""

		class_attributes = self.attributes().__class__
		base_directory = class_attributes.base_directory()
		index_html = class_attributes.index_html()

		html_filename = os.path.join(base_directory, index_html)
		with open(html_filename, 'w', encoding='utf-8') as a_file:
			self.write_header(a_file)
			self.write_body(a_file)
			self.write_footer(a_file)

	def write_body(self, file):
		"""ボディを書き出す。つまり、属性リストを書き出し、タプル群を書き出す。"""

		file.write("<tr>")
		file.write("\n")
	
		for attribute in self.table().attributes().names():
			#print(attribute)
			file.write("<td class=\"center-pink\"><strong>" + attribute + "</strong></td>")
			file.write("\n")
        
		file.write("<tr>")
		file.write("\n")

		file.write("<h1>"+self.attributes().caption_string()+"</h1>")
		file.write("\n")

		index = 0
		for a_tuple in self.table().tuples():	
			file.write("<tr>")
			file.write("\n")

			row_class = "even-row" if index % 2 == 0 else "odd-row"	
				
			for value in a_tuple.values():
				file.write(f"<td class=\"{row_class}\">")
				file.write(str(value))
				#print(str(value))
					
				file.write("</td>")
				file.write("\n")	
			file.write("</tr>")
			file.write("\n")
			index += 1
		(lambda x: x)(file) # NOP

	def write_footer(self, file):
		"""フッタを書き出す。"""

		file.write("</table>")
		file.write("\n")
		file.write("</body>")
		file.write("\n")
		file.write("</html>")
		file.write("\n")
		(lambda x: x)(file) # NOP

	def write_header(self, file):
		"""ヘッダを書き出す。"""

		file.write("<!DOCTYPE html>")
		file.write("\n")
		file.write("<html>")
		file.write("\n")
		file.write("<head>")
		file.write("<meta charset=\"UTF-8\">")
		file.write("<style>")
		file.write("table { border-collapse: collapse width: 100% }")
		file.write("th, td { border: 1px solid black; padding: 8px; text-align: left; }")
		file.write("th { background-color: #ffddbb; }")
		file.write(".even-row { background-color: #ddeeff; }")
		file.write(".odd-row { background-color: #ffffcc; }")
		file.write("img { display: block; margin: auto; }")
		file.write("</style>")
		file.write("</head>")
		file.write("\n")
		file.write("<body>")
		file.write("\n")
		file.write("<table>")
		file.write("\n")

		(lambda x: x)(file) # NOP
